src/pip/_internal/req/req_install.py

```diff
@@ -240,7 +247,7 @@ def supports_pyproject_editable(self) -> bool:
 
     @property
     def specifier(self) -> SpecifierSet:
-        return self.req.specifier
+        return self.checked_req.specifier
```

Comment:
A more appropriate fix would be

```python
if self.req is None:
    return SpecifierSet()
return self.req.specifier
```

Reply:
I'd rather that we have `assert self.req` here, making the existing contract that req needs to be non-None explicit.